app/src/main/java/com/itsaky/androidide/fragments/ProjectListFragment.kt

```diff
@@ -0,0 +1,140 @@
+import java.io.File
+
+
+class ProjectListFragment :
+  FragmentWithBinding<FragmentProjectListBinding>(R.layout.fragment_project_list,
+    FragmentProjectListBinding::bind) {
+
+  private var adapter: ProjectListAdapter? = null
+
+  private val viewModel by viewModels<MainViewModel>(ownerProducer = { requireActivity() })
+  private val projectInfoViewModel: ProjectInfoViewModel by viewModels { IDEViewModelProvider.Factory }
+
+  override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+    super.onViewCreated(view, savedInstanceState)
+
+    setupRecyclerView()
+    observeProjectInfoList()
+
+    projectInfoViewModel.getAllProjectInfo()
+
+    binding.exitButton.setOnClickListener {
+      viewModel.setScreen(MainViewModel.SCREEN_MAIN)
+    }
+
+  }
+
+  private fun setupRecyclerView() {
+    val layoutManager = LinearLayoutManager(requireContext())
+    layoutManager.orientation = LinearLayoutManager.VERTICAL
+    binding.list.layoutManager = layoutManager
+
+  }
+
+  private fun observeProjectInfoList() {
+    projectInfoViewModel.projectInfoList.observe(viewLifecycleOwner) {
+      val projectList = it.map { projectInfo ->
+        projectInfo.toProjectInfoDetails()
+      }
+
+      onFileListChanged(projectList)
+    }
+
+  }
+
+  private fun onFileListChanged(projectList: List<ProjectInfoDetails>) {
+    val sortOptions = ProjectInfoDetailsUtils.ProjectSortOptions(
+      ProjectInfoDetailsUtils.SortBy.OPEN_LAST, ProjectInfoDetailsUtils.Order.ASCENDING)
+      .createComparator()
+    val sortedProjectList = projectList.sortedWith(sortOptions)
+
+    adapter = ProjectListAdapter(sortedProjectList) { project, _ ->
+      openProject(project.file)
+    }
+
+    binding.list.adapter = adapter
+    setupNoProjectList(sortedProjectList.isEmpty())
+  }
+
+  private fun setupNoProjectList(isEmpty: Boolean) {
+    binding.list.visibility = if (isEmpty) View.GONE else View.VISIBLE
+    binding.noOpenProjectsSummary.visibility = if (isEmpty) View.VISIBLE else View.GONE
+
+    binding.noOpenProjectsSummary.movementMethod = LinkMovementMethod()
+    val openExistingProjectSpan: ClickableSpan = object : ClickableSpan() {
+      override fun onClick(widget: View) {
+        pickDirectory()
+      }
+    }
+
+    val sb = SpannableStringBuilder()
+    appendClickableSpan(sb, R.string.msg_empty_recent_projects, openExistingProjectSpan)
+    binding.noOpenProjectsSummary.text = sb
+  }
+
+  //Move this function to a utility class
+  private fun appendClickableSpan(
+    sb: SpannableStringBuilder,
+    @StringRes textRes: Int,
+    span: ClickableSpan,
+  ) {
+    val str = getString(textRes)
+    val split = str.split("@@", limit = 3)
+    if (split.size != 3) {
+      // Not a valid format
+      sb.append(str)
+      sb.append('\n')
+      return
+    }
+    sb.append(split[0])
+    sb.append(split[1], span, Spanned.SPAN_EXCLUSIVE_EXCLUSIVE)
+    sb.append(split[2])
+    sb.append('\n')
+  }
```

Comment:
Move this to a utility class and update the usage in editor activity as well.

Reply:
done.